pkg/lexer/lexer.go

```diff
@@ -0,0 +1,259 @@
+package lexer
+
+import (
+	"unicode"
+	"unicode/utf8"
+)
+
+type TokenType int
+
+// These are the main token types for shell
+const (
+	// Special tokens
+	ILLEGAL TokenType = iota
+	SEMICOLON
+	GREATER
+	LESS
+	GREATGREAT
+	LESSLESS
+	GREATAMPERSAND
+	LESSAMPERSAND
+	PIPE
+	AMPERSAND
+	IDENT
+	SINGLEQUOTE
+	DOUBLEQUOTE
+	BACKQUOTE
+	COMMENT
+	EOF
+)
+
+type Token struct {
+	Type TokenType
+	Val  string
+}
+
+// grammar
+// args -> IDENTS | empty
+// cmd and args -> IDENT args
+// pipe list -> PIPE cmd and args | cmd and args
+// io modifier -> GREAT IDENT | GREATGREAT IDENT | ...
+// io modifiers list -> io modifiers | empty
+// background optional -> ampersand | empty
+// cmd line -> pipe list io modifiers background optional NEWLINE | NEWLINE | ILLEGAL recovery.
+
+type Lexer struct {
+	input  string // input string
+	start  int    // start position of current token
```

Comment:
You could add a `ch` field of type `rune`, which will store the current rune.

Reply:
I tend to avoid the state manipulation, instead I use function returns which are more easy to understand IMO. This style of lexing is very much in line with the idea of state machine so a functional approach is more in line with the state machine aspect. But if you want I can change that as well.

Reply:
Using this approach will make sure that if we need to check what the current character is in a new state function, we can just read from the current state.

Reply:
I see no difference either way tbh. I prefer the returned one because that makes it explicit what is being returned instead of changing the ch attribute of the lexer struct. But its your call, let me know 😄

Reply:
I prefer it being a part of the state.